src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/ReadOnlySpanND.cs

```diff
@@ -0,0 +1,489 @@
+        [Obsolete("Equals() on ReadOnlySpanND will always throw an exception. Use the equality operator instead.")]
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public override bool Equals(object? obj) =>
+            throw new NotSupportedException(SR.NotSupported_CannotCallEqualsOnSpan);
+
+        /// <summary>
+        /// This method is not supported as spans cannot be boxed.
+        /// </summary>
+        /// <exception cref="NotSupportedException">
+        /// Always thrown by this method.
+        /// </exception>
+        [Obsolete("GetHashCode() on ReadOnlySpanND will always throw an exception.")]
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public override int GetHashCode() =>
+            throw new NotSupportedException(SR.NotSupported_CannotCallGetHashCodeOnSpan);
+#pragma warning restore CS0809
+
+        /// <summary>
+        /// Returns a 0-length read-only span whose base is the null pointer.
+        /// </summary>
+        public static ReadOnlySpanND<T> Empty => default;
+
+        /// <summary>
+        /// Casts a read-only span of <typeparamref name="TDerived"/> to a read-only span of <typeparamref name="T"/>.
+        /// </summary>
+        /// <typeparam name="TDerived">The element type of the source read-only span, which must be derived from <typeparamref name="T"/>.</typeparam>
+        /// <param name="items">The source read-only span. No copy is made.</param>
+        /// <returns>A read-only span with elements cast to the new type.</returns>
+        /// <remarks>This method uses a covariant cast, producing a read-only span that shares the same memory as the source. The relationships expressed in the type constraints ensure that the cast is a safe operation.</remarks>
+        public static ReadOnlySpanND<T> CastUp<TDerived>(ReadOnlySpanND<TDerived> items) where TDerived : class?, T
+        {
+            return new ReadOnlySpanND<T>(ref Unsafe.As<TDerived, T>(ref items._reference), items._lengths, items._strides, items._isPinned);
+        }
+
+        /// <summary>Gets an enumerator for this span.</summary>
+        public Enumerator GetEnumerator() => new Enumerator(this);
+
+        /// <summary>Enumerates the elements of a <see cref="ReadOnlySpan{T}"/>.</summary>
+        public ref struct Enumerator
+        {
+            /// <summary>The span being enumerated.</summary>
+            private readonly ReadOnlySpanND<T> _span;
+            private readonly Span<nint> _curIndices;
+            /// <summary>The total item count.</summary>
+            private nint _items;
+
+            /// <summary>Initialize the enumerator.</summary>
+            /// <param name="span">The span to enumerate.</param>
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            internal Enumerator(ReadOnlySpanND<T> span)
+            {
+                _span = span;
+                _items = -1;
+                _curIndices = new nint[_span.Rank];
```

Comment:
In the naive case where we always have a fully expanded tensor yes we can. We don't always have a fully expanded tensor though. We might be able to still do it with a single nint, but then we would have to do extra math for the conversion and I don't think it would save us much if anything.

Thant being said I'm happy to make the change if you think extra cpu would be better than that allocation. The allocation is small though as its just a number of elements equal to the number of dimensions, the overwhelming majority of the time that will be less than 4.

Reply:
> We might be able to still do it with a single nint, but then we would have to do extra math for the conversion and I don't think it would save us much if anything.

Is the math expensive? Allocating as part of creating a span or slicing a span is... unexpected.

Another option to consider would be prioritizing the most common sizes we expect, e.g. include an inline array of length 4 in all spans, in order to handle the first four indices, and anything beyond that requires allocation.

Reply:
There's going to be a balance here regardless. The multidimensional span is going to be large enough that we're likely going to need to pass it by `in` to avoid the expense of shadow copies (and will strictly be in "violation" of the FDG around structs not being bigger than 24 bytes). However, we really don't want it to be over 64 bytes in length, as that will take more than a single cache line per struct and cause its own problems.

Given that, I think the closest to an optimal setup, that avoids allocations for common sizes, is one of:
```csharp
public ref struct SpanND<T>
{
    private ref T _reference;       // 8 bytes
    private nint[]? _shape;         // 8 bytes

    private int _linearLength;      // 4 bytes
    private int _rank;              // 4 bytes

    private fixed int _lengths[5];  // 5x4 bytes (20)
    private fixed int _strides[5];  // 5x4 bytes (20)
}
```

-or-

public ref struct SpanND<T>
{
    private ref T _reference;       // 8 bytes
    private nint[]? _shape;         // 8 bytes

    private nint _linearLength;     // 8 bytes
    private int _rank;              // 4 bytes
    private int _flags;             // 4 bytes

    private fixed int _lengths[4];  // 4x4 bytes (16)
    private fixed int _strides[4];  // 4x4 bytes (16)
}

-or-

```csharp
public ref struct SpanND<T>
{
    private ref T _reference;       // 8 bytes
    private nint[]? _shape;         // 8 bytes

    private fixed nint _lengths[3]; // 3x8 bytes (24)
    private fixed nint _strides[3]; // 3x8 bytes (24)
}
```

With any of these setups, we would exactly fit in 64 bytes. We'd have the interior pointer, the single allocation used for the shape information for special circumstances (large spans, more than the inline dimension count, etc), and the lengths/stride information for up to a specific number of dimensions. We'd determine whether to use the inline buffers or the `_shape` field based on if the latter was `null` or not. If the `shape` was non-null, it'd be a single allocation representing the data in a particular order. -- This also means that `default` works as expected in all cases, which is nice.

In the first/second case, we'd make the lengths/strides for dimensions be `int` so they take up half the space, which should be enough for most types of multidimensional slices (unless some dimension is over 2.14 billion elements). We'd be able to explicitly track the rank (`int` is always sufficient here). For the first we'd track up to 5 dimensions inline and would accordingly use `int` for the linear length, limiting it to 2.14b elements (requiring `_shape` if its over this). The second deviates and instead allows `linearLength` to be an `nint` so it can always be tracked inline and then allows tracking a flags field for any additional metadata we may need/want. This limits the lengths/strides info to 4 dimensions, but that is likely to be overall more common and so the tradeoff is likely worth it.

In the third case we'd instead track the dimensions as `nint` so it works for large data as well and would only use `_shape` if there were more than 3 dimensions. We'd set a length/stride entry to `0` (or some sentinel) if it was unused and that'd give us a way to determine the rank (but this would be dynamic). The linear length would similarly need to be computed dynamically via `rank` multiplications for these small dimension counts, but that should be fine.

When `_shape` was non-null, it would end up being at least `_rank * 2` in length. Where in all cases it tracks `_rank` length entries, followed by `_rank` stride entries. In the case of the first, index 0 of the shape would be the `linearLength` and so `_shape` would be `_rank * 2 + 1` in length. In the case of the third we'd similarly have the `linearLength` and then explicitly have the `rank` information at index 1 (so `_shape` would be `_rank * 2 + 2` in length).

-- There are of course minor tweaks/deviations for this, and we could opt to split some info between `_shape` and the inline buffers for convenience/etc. But the general premise is to cover at least 3 dimensions (which will be the most common), to cover at least 2.14b elements per dimension (again the most common), and to avoid needing to do "too much" branching in the core path for handling any given multidimensional span.

Reply:
-- Notably this wasn't something that was being worried about for the prototype. The prototype was doing the naive thing so that it could get out in time for users to play around with the feature.

Reply:
So then do I we need to change anything for this pr?

And Stephen was also talking about the enumerator allocating. It currently allocates (and it won't let me do a stack alloc unless I'm doing something wrong) an array to help it iterate correctly.

Reply:
The same general premise would be used to optimize the struct enumerator for the common case.